drivers/mcp2515/candev_mcp2515.c

```diff
@@ -0,0 +1,570 @@
+                break;
+            }
+        }
+        break;
+    default:
+        res = -ENOTSUP;
+    }
+
+    return res;
+}
+
+static int _get(candev_t *candev, canopt_t opt, void *value, size_t max_len)
+{
+    candev_mcp2515_t *dev = (candev_mcp2515_t *)candev;
+    int res = 0;
+
+    DEBUG("Inside mcp2515 get opt=%d\n", opt);
+    switch (opt) {
+    case CANOPT_BITTIMING:
+        if (max_len < sizeof(candev->bittiming)) {
+            res = -EOVERFLOW;
+        }
+        else {
+            memcpy(value, &candev->bittiming, sizeof(candev->bittiming));
+            res = sizeof(candev->bittiming);
+        }
+        break;
+    case CANOPT_RX_FILTERS:      /**< rx filters */
+        if (max_len % sizeof(struct can_filter) != 0) {
+            res = -EOVERFLOW;
+        }
+        else {
+            res = 1;
+        }
+        break;
+    case CANOPT_BITTIMING_CONST:
+        if (max_len < sizeof(bittiming_const)) {
+            res = -EOVERFLOW;
+        }
+        else {
+            memcpy(value, &bittiming_const, sizeof(bittiming_const));
+            res = sizeof(bittiming_const);
+        }
+        break;
+    case CANOPT_CLOCK:
+        if (max_len < sizeof(uint32_t)) {
+            res = -EOVERFLOW;
+        }
+        else {
+            *((uint32_t *)value) = (uint32_t)(dev->conf->clk / 2);
+            res = sizeof(uint32_t);
+        }
+        break;
+    default:
+        res = -ENOTSUP;
+        break;
+    }
+
+    return res;
+}
+
+static int _set_filter(candev_t *dev, const struct can_filter *filter)
+{
+    DEBUG("inside _set_filter of MCP2515\n");
+    int filter_added = -1;
+    struct can_filter f = *filter;
+    int res = -1;
+    enum mcp2515_mode mode;
+
+    candev_mcp2515_t *dev_mcp = (candev_mcp2515_t *) dev;
+
+    if (f.can_mask == 0) {
+        return -EINVAL; /* invalid mask */
+    }
+
+    mode = mcp2515_get_mode(dev_mcp);
+    res = mcp2515_set_mode(dev_mcp, MODE_CONFIG);
+    if (res != MODE_CONFIG) {
+        return -1;
+    }
+
+    if ((f.can_id & CAN_EFF_FLAG) == CAN_EFF_FLAG) {
+        f.can_mask &= CAN_EFF_MASK;
+    }
+    else {
+        f.can_mask &= CAN_SFF_MASK;
+    }
+
+    /* Browse on each mailbox to find an empty space */
+    int mailbox_index = 0;
+    while (mailbox_index < MCP2515_RX_MAILBOXES && filter_added == -1) {
+        /* mask unused */
+        if (dev_mcp->masks[mailbox_index] == 0) {
+            /* set mask */
+            mcp2515_set_mask(dev_mcp, mailbox_index, f.can_mask);
+            /* set filter */
+            mcp2515_set_filter(dev_mcp, MCP2515_FILTERS_MB0 * mailbox_index, f.can_id);
+
+            /* save filter */
+            dev_mcp->masks[mailbox_index] = f.can_mask;
+            dev_mcp->filter_ids[mailbox_index][0] = f.can_id;
+
+            /* function succeeded */
+            filter_added = 1;
+        }
+
+        /* mask existed and same mask */
+        else if (dev_mcp->masks[mailbox_index] == f.can_mask) {
```

Comment:
could it happen that (and is it acceptable if)?:
- mailbox0 mask is 0
- mailbox1 mask is f.can_mask
- the new mask will be set into mailbox0 and the two mailboxes would have the same mask (bcs it is first tested for 0 and then for existing mask)